### ISSUE TITLE
hunspellLoader executing before emscripten module onRuntimeInitialized on high load arm systems

### ISSUE THREAD

**Issue by bluthen** [kwonoj/hunspell-asm#193]
When load hunspell on system is using a lot of memory and has high load I've been seeing the message:
`'FS' was not exported. add it to EXTRA_EXPORTED_RUNTIME_METHODS (see the FAQ)`
Hard coding FS right after FS inits in lib/hunspell.js I get:
`you need to wait for the runtime to be ready (e.g. wait for main() to be called)`

I've put some debug statements around. And I'm seeing hunspellLoader running before in emscripten-wasm-loader onRuntimeInitialized in constructModule gets called.

If I add a timeout long enough in the callback sent to getModuleLoader so onRuntimeInitialized gets called before calling hunspellLoader then it runs fine without error.

I would understand if this is not a priority, the system is super bogged down, and I can't replicated it besides on a heavily loaded arm machine (like an iPad running a way too bloated webapp). However this suggests to me there is an async issue somewhere that would be good to fix anyway.

I am looking at how emscripten-wasm-loader works however I am a bit confused on where it actually waits on the constructedModule promise.

**Comment by kwonoj** [comment]
Loading of wasm binary in emscripten-wasm-loader is fairly simple, 

first it runs loaded binary (https://github.com/kwonoj/emscripten-wasm-loader/blob/master/src/getModuleLoader.ts#L60-L69) and then callback to `hunspellLoader` https://github.com/kwonoj/hunspell-asm/blob/50fab4475fb9e61298c20c908f2d77a623dfcec4/src/loadModule.ts#L29

so I'd be quite surprised runtime environment can change order of execution honestly.

I guess FS is just red herring, hunspell-asm@3 no longer uses FS so it's expected to not exporte.

**Comment by kwonoj** [comment]
The only async execution part is `await asmModule.initializeRuntime` where we wait emscripten binary become ready, and if somehow execution occur before it is possible to see `you need to wait for the runtime to be ready (e.g. wait for main() to be called)` message. I can't say anything more unless if I have some reproducible way, but I'd doubt about race condition here for now.

**Comment by kwonoj** [comment]
What's order of logs meanwhile? curious if it shows some odd order.

**Comment by bluthen** [comment]
> ^^^^^^ hunspellLoader:  – Wed Jul 17 2019 17:23:15 GMT-0500 (CDT)   
> FS.mkdir is not a function. (In 'FS.mkdir(memPathId)', 'FS.mkdir' is undefined)
> ^^^^^^ hunspellLoader:  – Wed Jul 17 2019 17:23:15 GMT-0500 (CDT)   
> FS.mkdir is not a function. (In 'FS.mkdir(memPathId)', 'FS.mkdir' is undefined)
> !!!!!!!!!! constructModule onRuntimeInitialized: – Wed Jul 17 2019 17:23:19 GMT-0500 (CDT)   
> !!!!!!!!!! constructModule onRuntimeInitialized: – Wed Jul 17 2019 17:23:24 GMT-0500 (CDT)  

I am loading two instances of hunspell in this case with different dictionaries. I put logs at the top of hunspellLoader and in constructModules onRuntimeInitialized callback to output the messages.

This is why it seemed like getModuleLoader was not waiting for onRuntimeInitialized to call it's callback (the wrapper function that calls hunspellLoader).

I see the onRuntimeInitialized callback executing seconds after it tried to call hunspellLoader.

**Comment by kwonoj** [comment]
My q about log is internal log of hunspell-asm itself, can be wired via `enableLogger` fn.

**Comment by kwonoj** [comment]
and I presume this is latest 3.0.2?

**Comment by bluthen** [comment]
I will get back to you on all those, sorry if I filed a bad bug. I also thought of something else I might be doing wrong.

**Comment by kwonoj** [comment]
no worries at all.

**Comment by bluthen** [comment]
Yes about newest version.

Logging:
desktop (no problem)
```
hunspell::loadModule: loading hunspell wasm binary Object
hunspellLoader::loadModule: constructed module object for runtime
hunspellLoader::initializeRuntime: successfully initialized module
hunspellLoader::loadModule: initialized wasm binary Runtime
hunspellLoader: mount path for bufferFile created at /lj0Rp8ms8NcJ6bR2H0ojPuzzK5aHfo9gTaIhJwzEOBVKr
```

Loaded ipad:
```
hunspellLoader::loadModule: constructed module object for runtime
hunspellLoader::loadModule: initialized wasm binary Runtime
hunspellLoader::loadModule: failed to initialize wasm binary runtime
hunspellLoader::initializeRuntime: successfully initialized module
```

**Comment by bluthen** [comment]
That was it, sorry about filing a false bug report.

**Comment by bluthen** [comment]
Later I was thinking, maybe it should throw a exception indicating it timed out instead of trying to load and throwing a FS.mkdir undefined exception. This would help future people that ever run into this situation.

I think when we turned off emscripten optimization it made timing out a little bit more likely. I can try making a PR sometime next week if you think this is a good idea @kwonoj

**Comment by kwonoj** [comment]
Actually emscriptem-wasm-loader already throws:https://github.com/kwonoj/emscripten-wasm-loader/blob/master/src/getModuleLoader.ts#L72

Not sure why in you case it didn't bubbled up. If you can find fix for it, think we may add it.

**Comment by bluthen** [comment]
Would it be because the timeout just does resolve(false) instead of _reject(new Error('Timeout reached')); 
https://github.com/kwonoj/emscripten-wasm-loader/blob/master/src/constructModule.ts#L47

So then since it resolved and not rejected it run the factoryLoader, which in our case is the hunspellLoader and you get the FS.mkdir undefined issue from:
https://github.com/kwonoj/hunspell-asm/blob/master/src/hunspellLoader.ts#L24

I don't know for sure, as I'm still a little confused on how constructedModule eventually gets called.

**Comment by kwonoj** [comment]
v3.0.3 published.
